cache/disk/metrics.go

```diff
@@ -118,17 +126,97 @@ func (m *metricsDecorator) FindMissingCasBlobs(ctx context.Context, blobs []*pb.
 		"kind":   "cas",
 		"status": hitStatus,
 	}
+	m.addCategoryLabels(ctx, hitLabels)
 	hits := m.counter.With(hitLabels)
 
 	missLabels := prometheus.Labels{
 		"method": containsMethod,
 		"kind":   "cas",
 		"status": missStatus,
 	}
+	m.addCategoryLabels(ctx, missLabels)
 	misses := m.counter.With(missLabels)
 
 	hits.Add(float64(numFound))
 	misses.Add(float64(numMissing))
 
 	return digests, nil
 }
+
+func (m *metricsDecorator) Put(ctx context.Context, kind cache.EntryKind, hash string, size int64, r io.Reader) error {
+	err := m.diskCache.Put(ctx, kind, hash, size, r)
+	if err != nil {
+		return err
+	}
+
+	lbls := prometheus.Labels{"method": putMethod, "kind": kind.String(), "status": emptyStatus}
+	m.addCategoryLabels(ctx, lbls)
+	m.counter.With(lbls).Inc()
+
+	return nil
+}
+
+// Update prometheus labels based on HTTP and gRPC headers available via the context.
+func (m *metricsDecorator) addCategoryLabels(ctx context.Context, labels prometheus.Labels) {
+
+	if len(m.categories) == 0 {
+		return
+	}
+
+	httpHeaders := getHttpHeaders(ctx)
+	var grpcHeaders metadata.MD
+	if httpHeaders == nil {
+		grpcHeaders = getGrpcHeaders(ctx)
+	}
+
+	for categoryNameLowerCase, allowedValues := range m.categories {
+		// Lower case is canonical for gRPC headers and convention for prometheus.
+		var headerValue string = ""
+		if grpcHeaders != nil {
+			grpcHeaderValues := grpcHeaders[categoryNameLowerCase]
+			if len(grpcHeaderValues) > 0 {
+				// Pick the first header with matching name if multiple headers with same name
+				headerValue = grpcHeaderValues[0]
+			}
+		} else if httpHeaders != nil {
+			headerValue = httpHeaders.Get(categoryNameLowerCase)
+		}
+		if len(headerValue) == 0 {
+			labels[categoryNameLowerCase] = ""
+		} else if contains(allowedValues, headerValue) {
+			labels[categoryNameLowerCase] = headerValue
+		} else {
+			labels[categoryNameLowerCase] = "other"
+		}
+	}
+}
+
+func contains(s []string, e string) bool {
+	for _, a := range s {
+		if a == e {
+			return true
+		}
+	}
+	return false
+}
+
+type httpHeadersContextKey struct{}
+
+// Creates a context copy with HTTP headers attached.
+func ContextWithHttpHeaders(ctx context.Context, headers *http.Header) context.Context {
+	return context.WithValue(ctx, httpHeadersContextKey{}, headers)
+}
+
+// Retrieves HTTP headers from context. Minimizes type safety issues.
+func getHttpHeaders(ctx context.Context) *http.Header {
+	headers, ok := ctx.Value(httpHeadersContextKey{}).(*http.Header)
+	if !ok {
+		return nil
+	}
+	return headers
+}
+
+func getGrpcHeaders(ctx context.Context) metadata.MD {
+	grpcHeaders, _ := metadata.FromIncomingContext(ctx)
+	return grpcHeaders
+}
```

Comment:
I wonder if it would be a little cleaner to find http/grpc headers in the server package (separately in http and grpc code), and then add them as a single custom value in the context? Then we could simplify addCategoryLabels a little.